### PR TITLE
*: Support count aggregate function in new plan and add unit tests.

### PR DESCRIPTION
**Pull request by shenli**
@coocood @qiuyesuifeng

### CODE REVIEW

File: ast/functions.go [hunk 1]
```diff
@@ -382,3 +402,221 @@ func (n *AggregateFuncExpr) Accept(v Visitor) (Node, bool) {
 	}
 	return v.Leave(n)
 }
+
+// Update is used for update aggregate context.
```
Comment (qiuyesuifeng) [inline ast/functions.go]: updating?

File: ast/functions.go [hunk 2]
```diff
@@ -382,3 +402,221 @@ func (n *AggregateFuncExpr) Accept(v Visitor) (Node, bool) {
...
+	switch name {
+	case AggFuncCount:
+		return n.updateCount()
+	case AggFuncFirstRow:
+		return n.updateFirstRow()
+	}
+	return nil
+}
+
+// GetContext get aggregate evaluation context for the current group.
```
Comment (qiuyesuifeng) [inline ast/functions.go]: gets

File: ast/functions.go [hunk 3]
```diff
@@ -364,6 +379,11 @@ type AggregateFuncExpr struct {
 	// For example, column c1 values are "1", "2", "2",  "sum(c1)" is "5",
 	// but "sum(distinct c1)" is "3".
 	Distinct bool
+
+	CurrentGroup string
```
Comment (qiuyesuifeng) [inline ast/functions.go]: What is the meaning of `CurrentGroup`, could u show an example?

File: ast/functions.go [hunk 4]
```diff
@@ -364,6 +379,11 @@ type AggregateFuncExpr struct {
...
+	// ContextPerGroup is used to store aggregate evaluation context.
+	// Each entry for a group.
+	ContextPerGroup map[string](*AggEvaluateContext)
```
Comment (qiuyesuifeng) [inline ast/functions.go]: ContextPerGroup sounds like a single context, but the type is a map here.

File: ast/functions.go [hunk 5]
```diff
@@ -382,3 +402,221 @@ func (n *AggregateFuncExpr) Accept(v Visitor) (Node, bool) {
...
+
+func (n *AggregateFuncExpr) updateCount() error {
+	ctx := n.GetContext()
+	var value interface{}
+	if len(n.Args) > 0 {
+		hasNil := false
+		for _, a := range n.Args {
+			value = a.GetValue()
+			if value == nil {
+				hasNil = true
```
Comment (qiuyesuifeng) [inline ast/functions.go]: Can directly return?

File: ast/functions.go [hunk 6]
```diff
@@ -382,3 +402,221 @@ func (n *AggregateFuncExpr) Accept(v Visitor) (Node, bool) {
...
+		if a.detecting {
+			// Enter a new context, skip it.
+			// For example: select sum(c) + c + exists(select c from t) from t;
+			return n, true
+		}
+	}
+	return n, a.HasAggFunc
+}
+
+// Leave implemets Visitor interface.
```
Comment (qiuyesuifeng) [inline ast/functions.go]: implements

File: ast/functions.go [hunk 7]
```diff
@@ -382,3 +402,221 @@ func (n *AggregateFuncExpr) Accept(v Visitor) (Node, bool) {
...
+func (a *AggFuncDetector) Leave(n Node) (node Node, ok bool) {
+	return n, !a.HasAggFunc
+}
+
+// AggregateFuncExtractor visit Expr tree.
```
Comment (qiuyesuifeng) [inline ast/functions.go]: visits

File: ast/functions.go [hunk 8]
```diff
@@ -382,3 +402,221 @@ func (n *AggregateFuncExpr) Accept(v Visitor) (Node, bool) {
...
+// GetContext get aggregate evaluation context for the current group.
+// If it is nil, add a new context into ContextPerGroup.
+func (n *AggregateFuncExpr) GetContext() *AggEvaluateContext {
+	if n.ContextPerGroup == nil {
+		n.ContextPerGroup = make(map[string](*AggEvaluateContext))
+	}
+	if _, ok := n.ContextPerGroup[n.CurrentGroup]; !ok {
+		c := &AggEvaluateContext{}
+		if n.Distinct {
+			c.Distinct = CreateAggregateDistinct(n.F, n.Distinct)
```
Comment (qiuyesuifeng) [inline ast/functions.go]: CreateAggregateDistinct function has inner checked Distinct value, so seems needless for  n.Distinct  check.

File: ast/functions.go [hunk 9]
```diff
@@ -382,3 +402,221 @@ func (n *AggregateFuncExpr) Accept(v Visitor) (Node, bool) {
...
+func (a *AggregateDistinct) IsDistinct(v ...interface{}) (bool, error) {
+	// no distinct flag
+	if a.Distinct == nil {
+		return true, nil
+	}
+
+	k := v
+	r, err := a.Distinct.Get(k)
+	if err != nil {
+		return false, nil
```
Comment (qiuyesuifeng) [inline ast/functions.go]: return `false, errors.Trace(err)`

File: ast/functions.go [hunk 10]
```diff
@@ -382,3 +402,221 @@ func (n *AggregateFuncExpr) Accept(v Visitor) (Node, bool) {
...
+	}
+
+	if len(r) > 0 {
+		// we save a same value before
+		return false, nil
+	}
+
+	if err := a.Distinct.Set(k, []interface{}{true}); err != nil {
+		return false, err
```
Comment (qiuyesuifeng) [inline ast/functions.go]: return false, errors.Trace(err)

File: ast/functions.go [hunk 11]
```diff
@@ -382,3 +402,221 @@ func (n *AggregateFuncExpr) Accept(v Visitor) (Node, bool) {
...
+func (a *AggregateDistinct) Clear() {
+	if a.Distinct == nil {
+		return
+	}
+
+	// drop does nothing, no need to check error
+	a.Distinct.Drop()
+	// CreateTemp returns no error, no need to check error
+	// later we may use another better way instead of memkv
+	a.Distinct, _ = memkv.CreateTemp(true)
```
Comment (qiuyesuifeng) [inline ast/functions.go]: Why here create a new temp map? If so, i think function name should be Reset.

File: ast/functions.go [hunk 12]
```diff
@@ -382,3 +402,221 @@ func (n *AggregateFuncExpr) Accept(v Visitor) (Node, bool) {
...
+// Leave implemets Visitor interface.
+func (a *AggFuncDetector) Leave(n Node) (node Node, ok bool) {
+	return n, !a.HasAggFunc
+}
+
+// AggregateFuncExtractor visit Expr tree.
+// It converts ColunmNameExpr to AggregateFuncExpr and collects AggregateFuncExpr.
+type AggregateFuncExtractor struct {
+	inAggregateFuncExpr bool
+	inSelectFields      bool
```
Comment (qiuyesuifeng) [inline ast/functions.go]: Seems inSelectFields has been used?

File: ast/functions.go [hunk 13]
```diff
@@ -382,3 +402,221 @@ func (n *AggregateFuncExpr) Accept(v Visitor) (Node, bool) {
...
+	// AggFuncs is the collected AggregateFuncExprs.
+	AggFuncs  []*AggregateFuncExpr
+	detecting bool
+}
+
+// Enter implemets Visitor interface.
+func (a *AggregateFuncExtractor) Enter(n Node) (node Node, skipChildren bool) {
+	defer func() {
+		a.detecting = true
```
Comment (qiuyesuifeng) [inline ast/functions.go]: It is not necessary to use a defer, only set `a.detecting = true` before return is enough.

File: ast/functions.go [hunk 14]
```diff
@@ -382,3 +402,221 @@ func (n *AggregateFuncExpr) Accept(v Visitor) (Node, bool) {
...
+			// The c after plus should be evaluated only once.
+			agg := &AggregateFuncExpr{
+				F:    AggFuncFirstRow,
+				Args: []ExprNode{n.(ExprNode)},
+			}
+			a.AggFuncs = append(a.AggFuncs, agg)
+			return agg, true
+		}
+	}
+	return n, true
```
Comment (qiuyesuifeng) [inline ast/functions.go]: Should a.detecting be set false in Leave function?

File: ast/functions.go [hunk 15]
```diff
@@ -382,3 +402,221 @@ func (n *AggregateFuncExpr) Accept(v Visitor) (Node, bool) {
...
+}
+
+// AggregateDistinct handles distinct data for aggregate function: count, sum, avg, and group_concat.
+type AggregateDistinct struct {
+	// Distinct is a memory key-value map.
+	// Now we have to use memkv Temp, later may be use map directly
+	Distinct memkv.Temp
```
Comment (qiuyesuifeng) [inline ast/functions.go]: Distinct looks like a boolean variable, how about tempTable or tempMap?

File: executor/executor.go [hunk 16]
```diff
@@ -36,6 +36,7 @@ var (
 	_ Executor = &IndexScanExec{}
 	_ Executor = &LimitExec{}
 	_ Executor = &SelectFieldsExec{}
+	_ Executor = &AggregateExec{}
```
Comment (qiuyesuifeng) [inline executor/executor.go]: Please sort in alphabetical order.

File: ast/functions.go (hunk 3 above)
Comment (shenli) [inline ast/functions.go]: When call update or evaluate, we should specify the group key by setting this attribute.

File: ast/functions.go (hunk 4 above)
Comment (shenli) [inline ast/functions.go]: It means one context for a group. The "Per" means for every group.

File: ast/functions.go (hunk 1 above)
Comment (shenli) [inline ast/functions.go]: "Update" is the function name. We need a verb for this function.

File: ast/functions.go (hunk 14 above)
Comment (shenli) [inline ast/functions.go]: No

File: ast/functions.go (hunk 15 above)
Comment (shenli) [inline ast/functions.go]: I will replace it with a map. It is copied from the old code.

File: ast/functions.go (hunk 9 above)
Comment (shenli) [inline ast/functions.go]: AggregateDistinct will be replaced with a map in the next pr.

**Comment by shenli** [comment]
PTAL @qiuyesuifeng

File: ast/functions.go (hunk 4 above)
Comment (coocood) [inline ast/functions.go]: Agree with @qiuyesuifeng, consider `Contexts` or `ContextMap`.

File: ast/functions.go [hunk 17]
```diff
@@ -382,3 +402,219 @@ func (n *AggregateFuncExpr) Accept(v Visitor) (Node, bool) {
...
+// Update is used for update aggregate context.
+func (n *AggregateFuncExpr) Update() error {
+	name := strings.ToLower(n.F)
+	switch name {
+	case AggFuncCount:
+		return n.updateCount()
+	case AggFuncFirstRow:
+		return n.updateFirstRow()
+	}
+	return nil
```
Comment (coocood) [inline ast/functions.go]: If the name doesn't match any function name, we return an error.

File: ast/functions.go [hunk 18]
```diff
@@ -382,3 +402,219 @@ func (n *AggregateFuncExpr) Accept(v Visitor) (Node, bool) {
...
+
+// AggFuncDetector visits Expr tree to check if it contains AggregateFuncExpr.
+type AggFuncDetector struct {
+	HasAggFunc bool
+	detecting  bool
+}
+
+// Enter implements Visitor interface.
+func (a *AggFuncDetector) Enter(n Node) (node Node, skipChildren bool) {
+	defer func() {
```
Comment (coocood) [inline ast/functions.go]: defer is too expensive here in `Enter` because it will be called many times during visiting.

File: ast/functions.go [hunk 19]
```diff
@@ -382,3 +402,219 @@ func (n *AggregateFuncExpr) Accept(v Visitor) (Node, bool) {
...
+type AggregateFuncExtractor struct {
+	inAggregateFuncExpr bool
+	// AggFuncs is the collected AggregateFuncExprs.
+	AggFuncs  []*AggregateFuncExpr
+	detecting bool
+}
+
+// Enter implements Visitor interface.
+func (a *AggregateFuncExtractor) Enter(n Node) (node Node, skipChildren bool) {
+	defer func() {
```
Comment (coocood) [inline ast/functions.go]: And here, remove defer.

File: ast/functions.go [hunk 20]
```diff
@@ -382,3 +402,219 @@ func (n *AggregateFuncExpr) Accept(v Visitor) (Node, bool) {
...
+		a.detecting = true
+	}()
+	switch v := n.(type) {
+	case *AggregateFuncExpr:
+		a.inAggregateFuncExpr = true
+		a.AggFuncs = append(a.AggFuncs, v)
```
Comment (coocood) [inline ast/functions.go]: I think this line should be moved to `Leave`, all AggFunc will be appended in one place.

File: executor/executor.go [hunk 21]
```diff
@@ -693,3 +694,107 @@ func (e *SortExec) Next() (*Row, error) {
 func (e *SortExec) Close() error {
 	return e.Src.Close()
 }
+
+// For select stmt with aggregate function but without groupby clasue,
+// We consider there is a single group with key singleGroup.
+const singleGroup = "SingleGroup"
+
+// AggregateExec deal with all the aggregate functions.
+// It is built from Aggregate Plan. When Next() is called, it read all the data from Src and update all the items in AggFuncs.
```
Comment (coocood) [inline executor/executor.go]: reads, updates.

File: executor/executor.go [hunk 22]
```diff
@@ -693,3 +694,107 @@ func (e *SortExec) Next() (*Row, error) {
...
+		if e.executed {
+			return false, nil
+		}
+	}
+	e.executed = true
+	// TODO: evaluate groupby key and set currentgroup to each aggfunc.
+	groupKey := singleGroup
+	e.groupMap[groupKey] = true
+	for _, af := range e.AggFuncs {
+		if len(af.Args) > 0 {
```
Comment (coocood) [inline executor/executor.go]: This line can be removed.

**Comment by shenli** [comment]
PTAL @coocood @qiuyesuifeng

File: ast/functions.go (hunk 17 above)
Comment (shenli) [inline ast/functions.go]: It will be checked at validator and support checker.

**Comment by coocood** [comment]
LGTM

File: ast/functions.go [hunk 23]
```diff
@@ -382,3 +402,211 @@ func (n *AggregateFuncExpr) Accept(v Visitor) (Node, bool) {
...
+		}
+		n.contextPerGroupMap[n.CurrentGroup] = c
+	}
+	return n.contextPerGroupMap[n.CurrentGroup]
+}
+
+func (n *AggregateFuncExpr) updateCount() error {
+	ctx := n.GetContext()
+	var value interface{}
+	if len(n.Args) > 0 {
```
Comment (qiuyesuifeng) [inline ast/functions.go]: `len(n.Args) > 0` can be removed?

**Comment by qiuyesuifeng** [comment]
LGTM

**Pull request merged**